src/functions/addKvRecords.ts

```diff
@@ -0,0 +1,133 @@
+import { decResLengths } from '../constants';
+import { decryptResponse, encryptRequest, request } from '../shared/functions';
+import {
+  validateFwConstants,
+  validateKvRecord,
+  validateKvRecords,
+  validateSharedSecret,
+  validateUrl,
+} from '../shared/validators';
+
+/**
+ * `addKvRecords` takes in a set of key-value records and sends a request to add them to the
+ * Lattice.
+ * @category Lattice
+ * @returns A callback with an error or null.
+ */
+export async function addKvRecords ({
+  type = 0,
+  records,
+  caseSensitive = false,
+  client,
+}: AddKvRecordsRequestFunctionParams): Promise<Buffer> {
+  const { url, sharedSecret, fwConstants, validRecords } = validateAddKvRequest(
+    {
+      url: client.url,
+      fwConstants: client.getFwConstants(),
+      sharedSecret: client.sharedSecret,
+      records,
+    },
+  );
+
+  const payload = encodeAddKvRecordsRequest({
+    records: validRecords,
+    fwConstants,
+    type,
+    caseSensitive,
+  });
+
+  const encryptedPayload = encryptAddKvRecordsRequest({
+    payload,
+    sharedSecret,
+  });
+
+  const encryptedResponse = await requestAddKvRecords(encryptedPayload, url);
+
+  const { decryptedData, newEphemeralPub } = decryptAddKvRecordsResponse(
+    encryptedResponse,
+    sharedSecret,
+  );
+
+  client.ephemeralPub = newEphemeralPub;
+
+  return decryptedData;
+}
+
+export const validateAddKvRequest = ({
+  url,
+  fwConstants,
+  sharedSecret,
+  records,
+}: ValidateAddKvRequestParams) => {
+  const validUrl = validateUrl(url);
+  const validFwConstants = validateFwConstants(fwConstants);
+  const validSharedSecret = validateSharedSecret(sharedSecret);
+  const validRecords = validateKvRecords(records, validFwConstants);
+
+  return {
+    url: validUrl,
+    fwConstants: validFwConstants,
+    sharedSecret: validSharedSecret,
+    validRecords,
+  };
+};
+
+export const encodeAddKvRecordsRequest = ({
```

Comment:
Do you need to export encoders?

Reply:
Coming back after reviewing the whole code base. I think you export everything for testing purposes, right?

Reply:
Right--just for the tests.